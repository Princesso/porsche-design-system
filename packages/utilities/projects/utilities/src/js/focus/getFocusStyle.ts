import type { Theme } from '../theme';
import { themeDarkStateFocus, themeLightStateFocus } from '../theme';
import { borderRadiusMedium, borderRadiusSmall, borderWidthBase } from '../border';

type Size = 'small' | 'medium';
type Inset = Size;
type BorderRadius = Size;
type Options = {
  inset?: Inset | string;
  borderRadius?: BorderRadius | string;
  theme?: Theme;
};

export const getFocusStyle = (opts?: Options) => {
  const inset = opts?.inset === 'small' ? '-4px' : opts?.inset === 'medium' ? '-6px' : opts?.inset || '-4px';
  const borderRadius =
    opts?.borderRadius === 'small'
      ? borderRadiusSmall
      : opts?.borderRadius === 'medium'
      ? borderRadiusMedium
      : opts?.borderRadius || borderRadiusSmall;

  return {
    position: 'relative',
    // we can't use outline, since border-radius has no effect on outline style in Safari
    outline: 0,
    '&:focus::after': {
      content: '""',
      position: 'absolute',
      inset,
      pointerEvents: 'none', // necessary to be able to select elements behind pseudo-element
      border: `${borderWidthBase} solid ${opts?.theme === 'dark' ? themeDarkStateFocus : themeLightStateFocus}`,
      borderRadius,
    },
    // why? have a look at this article https://developer.paciellogroup.com/blog/2018/03/focus-visible-and-backwards-compatibility/
    '&:focus:not(:focus-visible)::after': {
      borderColor: 'transparent',
    },
  } as const;
};
