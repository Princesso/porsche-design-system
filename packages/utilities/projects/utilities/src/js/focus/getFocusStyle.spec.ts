import { getFocusStyle } from './getFocusStyle';
import * as fromFocus from './';

it('should provide all exports', () => {
  expect(Object.keys(fromFocus).length).toBe(1);
});

describe('getFocusStyle()', () => {
  it.each<Parameters<typeof getFocusStyle>>([[], [{ offset: '5rem' }], [{ theme: 'dark' }]])(
    'should return correct css for opts: %s',
    (...args) => {
      expect(getFocusStyle(...args)).toMatchSnapshot();
    }
  );
});
