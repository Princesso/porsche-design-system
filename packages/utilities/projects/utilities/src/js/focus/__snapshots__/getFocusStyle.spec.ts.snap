// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFocusStyle() should return correct css for opts: %s 1`] = `
{
  "&:focus::before": {
    "border": "2px solid #0A0AFF",
    "borderRadius": "4px",
    "bottom": "-4px",
    "content": """",
    "left": "-4px",
    "position": "absolute",
    "right": "-4px",
    "top": "-4px",
  },
  "&:focus:not(:focus-visible)::before": {
    "borderColor": "transparent",
  },
  "outline": 0,
  "position": "relative",
}
`;

exports[`getFocusStyle() should return correct css for opts: { offset: '5rem' } 1`] = `
{
  "&:focus::before": {
    "border": "2px solid #0A0AFF",
    "borderRadius": "4px",
    "bottom": "5rem",
    "content": """",
    "left": "5rem",
    "position": "absolute",
    "right": "5rem",
    "top": "5rem",
  },
  "&:focus:not(:focus-visible)::before": {
    "borderColor": "transparent",
  },
  "outline": 0,
  "position": "relative",
}
`;

exports[`getFocusStyle() should return correct css for opts: { theme: 'dark' } 1`] = `
{
  "&:focus::before": {
    "border": "2px solid #0A0AFF",
    "borderRadius": "4px",
    "bottom": "-4px",
    "content": """",
    "left": "-4px",
    "position": "absolute",
    "right": "-4px",
    "top": "-4px",
  },
  "&:focus:not(:focus-visible)::before": {
    "borderColor": "transparent",
  },
  "outline": 0,
  "position": "relative",
}
`;
